ztrouy/loncotes-library
# Basic Features

### Purpose
To enable users to interact with Materials, Patrons, and Checkouts in a basic manner

### Files Changed
- Setup migration in `20240506153447_InitialCreate.Designer.cs`, `20240506153447_InitialCreate.cs`, and `LoncotesLibraryDbContextModelSnapshot.cs`
- Added List property for Checkouts in `Patron.cs`, `PatronDTO.cs`, `Checkout.cs`, and `Material.cs`
- Created initial DTO class used for submitting a Checkout to create in new `CheckoutCreateDTO.cs`
- Created initial DTO class used for submitting a Material to create in new `MaterialCreateDTO.cs`
- Created initial DTO class used for returning a Material with all related Checkouts in new `MaterialDetailedDTO.cs`
- Created initial DTO class used for updating a Patron in new `PatronUpdateDTO.cs`
- Added all current endpoints in `Program.cs`
- Removed unnecessary whitespace in `LoncotesLibraryDbContex.cs`

### To Test
Fetch, and run the following commands in order
```
dotnet ef migrations add InitialCreate
```
```
dotnet ef database update
```
```
dotnet watch run
```
Then confirm the following
- GET endpoint `api/materials` returns all Materials, along with their Material Type and Genre
- GET endpoint `api/materials/{id}` returns the appropriate Material, with it's Material Type, Genre, all related Checkouts, and the Patron for each Checkout
- GET endpoint `api/materialtypes` returns all Material Types
- GET endpoint `api/genres` returns all Genres
- GET endpoint `api/patrons` returns all Patrons
- GET endpoint `api/patrons/{id}` returns the appropriate Patron with their Checkouts, and the Material and Material Type for each Checkout
- PUT endpoint `api/patrons/{id}` updates the Address and Email of a Patron when provided a valid request
- PUT endpoint `api/patrons/{id}/deactivate` sets the "IsActive" property of the appropriate Patron to "False"
- POST endpoint `api/checkouts` creates a new Checkout when provided a valid request
- PUT endpoint `api/checkouts/{id}/return` sets the "ReturnDate" property of the appropriate Checkout to the current Date Time
